### ISSUE TITLE
Revisiting Small Batch Training for Deep Neural Networks

### ISSUE THREAD

**Issue by chullhwan-song** [chullhwan-song/Reading-Paper#12]
https://arxiv.org/abs/1804.07612

**Comment by chullhwan-song** [comment]
### Abstract
*  이 논문은 survey 격 논문과 batch normalization에 대한 실험적 성격을 지님
*  결과적으로  우리가 알고있는 batch size(m)를 더 줄어야한다. 란 소리를 하고 있음  
    * bast performance: 2<= m<=32 일때라고 주장.  - **The best performance has been consistently obtained for mini-batch sizes between m = 2 and m = 32, which contrasts with recent work advocating the use of mini-batch sizes in the thousands.**
    * 참고로 나 자신한테도 당황스런 경험인데 특정 domain 인식에서 batch size를 줄였더니, 성능향상이 있어다는걸 경험.  따라서 이 논문은 나한테 어느정도 이에 대한 refer로써.. 
### background theory 
* SGD
* batch normalization 
### BATCH TRAINING PERFORMANCE
* hyper-param - batch size & learning rate
* 다양한 benchmark dataset : CIFAR-10, CIFAR-100 (Krizhevsky, 2009) and ImageNet (Krizhevsky et al., 2012). 
* cnn 모델 
     * CIFAR-10 and CIFAR-100 경우 :  AlexNet & ResNet 
        * BN, data augmentation and network depth 실험
#### IMPLEMENTATION DETAILS
* SGD optimization
* ![image](https://user-images.githubusercontent.com/40360823/42300026-50b789c4-8049-11e8-8eea-c657e56255a7.png)
* all learning rates and all batch sizes > performed over the **same number of epochs**,
* CIFAR-10 and CIFAR-100 datasets
    * AlexNet, ResNet-8, ResNet-20 and ResNet-32 models
       * The **reduced AlexNet** implementation uses convolutional layers with stride equal to 1, kernel sizes equal to [5; 3; 3; 3; 3], number of channels per layer equal to [16; 48; 96; 64; 64], max pool layers with 3x3 kernels and stride 2, and 256 hidden nodes per fully-connected layer. 
    * epochs : CIFAR-10  = 82, CIFAR-100 = 164
    * learning rate reduced : reduction by a factor of 10 at 50% and at 75% of the total number of  iterations.
* 밑에 부터는 살짝 두서없이.ㅠ
* CIFAR datasets
    * 50,000 example training, 10,000 example test
    * without augmentation of the training data vs pads the images by 4 pixels on each side,
takes a random 32x32 crop, and then randomly applies a horizontal flip.
* ImageNet - ResNet-50 model
* CIFAR experiments - 90 epochs, with a learning rate reduction by a factor of 10 at 30, 60 and 80 epochs,

#### PERFORMANCE WITHOUT BATCH NORMALIZATION
* CIFAR-10 test performance of the reduced AlexNet, ResNet-8 and ResNet-20 models
   *  without BN or data augmentation
   * best test accuracy is obtained for batch sizes m = 8 or smaller, while for the ResNet models the best accuracy is obtained for batch sizes m = 4 and m = 2.
![image](https://user-images.githubusercontent.com/40360823/42300057-7ef2e7f2-8049-11e8-8a6d-d13d0df2c9c4.png)
![image](https://user-images.githubusercontent.com/40360823/42300059-8207e4f6-8049-11e8-9afc-ec57f3780592.png)
* best test performance 76%
   * successful only for the cases of batch size m = 2, m = 4 and m = 8,

####  PERFORMANCE WITH BATCH NORMALIZATION
* convergence curves for CIFAR-10, ResNet-32 training with BN, for different batch sizes.
![image](https://user-images.githubusercontent.com/40360823/42300116-c58a1d84-8049-11e8-9ed2-c7e66620363a.png)
   *  4<=m<=64 일경우 best performance
   * learning rate는 고정시킴
* CIFAR-10 test performance of the AlexNet model with BN but without data augmentation, for different values of batch size and base learning rate.
![image](https://user-images.githubusercontent.com/40360823/42300121-cfb34952-8049-11e8-9b0a-62013cb2c36d.png)
   * Figure 5 : BN is commonly applied to the activations in all convolutional and fully-connected layers. 
       * best performance : 8<=m<=64
   * Figure 6 : only convolutional layers
       * best performance : 4<=m<=32     
* the training performance for CIFAR-10 (with and without data augmentation) and CIFAR-100 (with data augmentation) on the ResNet-32 model with BN, for different values of batch size and base learning rate.
![image](https://user-images.githubusercontent.com/40360823/42300145-e3163e28-8049-11e8-8891-13c5916974e1.png)
![image](https://user-images.githubusercontent.com/40360823/42300147-e5c43d8c-8049-11e8-8976-212edd3b0a8c.png)
    * best performance 
       * batch sizes m = 16 or smaller for CIFAR-10, and m = 8 or smaller for CIFAR-100.
* 이 이후에서 비슷한 실험을 했는데 일관되게 매우 작은 batch size를 가지고 학습했을때 best performance를 얻었다.
![image](https://user-images.githubusercontent.com/40360823/42300153-f2b25380-8049-11e8-8e2f-9fb6c976091a.png)
* 같이 읽어볼만한 논문 : https://arxiv.org/abs/1609.04836

**Comment by chullhwan-song** [comment]
https://blog.lunit.io/2018/08/03/batch-size-in-deep-learning/ 이 블로그가 훨씬 자세한 설명
